rust/src/handshake.rs

```diff
@@ -0,0 +1,308 @@
+
+// Author: Sascha Steinbiss <sascha@steinbiss.name>
+
+*/
+
+use libc::c_uchar;
+use std::os::raw::c_char;
+use tls_parser::{TlsCipherSuiteID, TlsExtensionType, TlsVersion};
+
+#[derive(Debug, PartialEq)]
+pub struct HandshakeParams {
+    pub(crate) tls_version: Option<TlsVersion>,
+    pub(crate) ciphersuites: Vec<TlsCipherSuiteID>,
+    pub(crate) extensions: Vec<TlsExtensionType>,
+    pub(crate) signature_algorithms: Vec<u16>,
+    pub(crate) domain: bool,
+    pub(crate) alpns: Vec<Vec<u8>>,
+    pub(crate) quic: bool,
+}
+
+impl Default for HandshakeParams {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl HandshakeParams {
+    #[inline]
+    pub(crate) fn is_grease(val: u16) -> bool {
+        match val {
+            0x0a0a | 0x1a1a | 0x2a2a | 0x3a3a | 0x4a4a | 0x5a5a | 0x6a6a | 0x7a7a | 0x8a8a
+            | 0x9a9a | 0xaaaa | 0xbaba | 0xcaca | 0xdada | 0xeaea | 0xfafa => true,
+            _ => false,
+        }
+    }
+
+    fn new() -> Self {
+        Self {
+            tls_version: None,
+            ciphersuites: Vec::with_capacity(20),
+            extensions: Vec::with_capacity(20),
+            signature_algorithms: Vec::with_capacity(20),
+            domain: false,
+            alpns: Vec::with_capacity(4),
+            quic: false,
+        }
+    }
+
+    pub(crate) fn set_tls_version(&mut self, version: TlsVersion) {
+        if Self::is_grease(u16::from(version)) {
+            return;
+        }
+        // Track maximum of seen TLS versions
+        match self.tls_version {
+            None => {
+                self.tls_version = Some(version);
+            }
+            Some(cur_version) => {
+                if u16::from(version) > u16::from(cur_version) {
+                    self.tls_version = Some(version);
+                }
+            }
+        }
+    }
+
+    pub fn add_alpn(&mut self, alpn: &[u8]) {
+        if alpn.is_empty() {
+            return;
+        }
+        self.alpns.push(alpn.to_vec());
+    }
+
+    pub(crate) fn add_cipher_suite(&mut self, cipher: TlsCipherSuiteID) {
+        if Self::is_grease(u16::from(cipher)) {
+            return;
+        }
+        self.ciphersuites.push(cipher);
+    }
+
+    pub(crate) fn add_extension(&mut self, ext: TlsExtensionType) {
+        if Self::is_grease(u16::from(ext)) {
+            return;
+        }
+        if ext == TlsExtensionType::ServerName {
+            self.domain = true;
+        }
+        self.extensions.push(ext);
+    }
+
+    pub(crate) fn add_signature_algorithm(&mut self, sigalgo: u16) {
+        if Self::is_grease(sigalgo) {
+            return;
+        }
+        self.signature_algorithms.push(sigalgo);
+    }
+}
+
+// Objects used to allow C to call into this struct via the below C ABI
+// that enables the return of a u8 pointer and length
+#[repr(C)]
+pub struct CStringData {
+    data: *const u8,
+    len: usize,
+}
+
+#[no_mangle]
+pub extern "C" fn SCHandshakeNew() -> *mut HandshakeParams {
```

Comment:
This naming prefix is quite generic. Should it be SCTLSHandshake? Or perhaps something like SCCryptoHandshake?

Reply:
Yeah I can make it either of these! Whichever one you prefer? There's no point me making another PR for this change only. I'd rather the whole team review it. Once everyone has reviewed I will make all appropriate changes as once!

Reply:
`SCTLSHandshake` makes sense to me!

Reply:
Lets go with `SCTLSHandshake`